---
title: 'NxtIRF: 3 - Data Visualisation'
author: "Alex Chit Hei Wong"
date: "`r Sys.Date()`"
output:
    rmarkdown::html_vignette:
        toc: true
vignette: >
    %\VignetteIndexEntry{NxtIRF: 3 - Data Visualisation}
    %\VignetteEngine{knitr::rmarkdown}
    \usepackage[utf8]{inputenc}
---

In previous vignettes, we showed how to create a NxtIRF reference and use it to analyse a sample dataset.
In this vignette, we will use the reference, the results of the differential expression (saved as an RDS)
file in the previous vignette, and the COV files to illustrate IR events in this dataset.

First, load the NxtIRF package:

```{r, eval = FALSE}
library(NxtIRF)
```

The following vignette will alternate between the GUI and CLI interfaces to show how equivalent steps
are performed with these two alternate user interfaces.

# Loading the Data

First navigate to the root directory under which we saved the reference, compiled experiment and differential
expression RDS file in the prior vignettes. Then open the NxtIRF GUI as shown:

## Graphics user interface:

```{r, eval = FALSE}
nxtIRF()
```

Navigate to the Experiment tab, and set the reference path (click the pink "Reference" button,
followed by the "Select reference path" button, and use the dialog box to select the directory
containing the reference. Next set the directory by clicking the pink "Construct Experiment" button, 
followed by "Choose NxtIRF FST output path". After selecting the directory where we constructed
the experiment, your NxtIRF app should look like this:

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Loading the Reference and Experiment"}
knitr::include_graphics("img/Vis_1_intro.png")
```

Now, click the "Build SummarizedExperiment" button inside the "Construct Experiment" box. The
SummarizedExperiment will be loaded when the "SummarizedExperiment Object" status box
(located above the pink "Reference" button) turns from yellow to green.

## Command line interface:

Load the SummarizedExperiment object using the steps similar to the Experiment vignette
(we assume you have set the working directory to the root directory):

```{r, eval = FALSE}
setwd("/path/to/project")
expr.df = Find_IRFinder_Output("./IRFinder_output")

colData = data.frame(sample = expr.df$sample)
colData$Treatment = data.table::tstrsplit(colData$sample, split="_")[[1]]
View(colData)
```

Now, we use `MakeSE()` to assemble the SummarizedExperiment.

```{r, eval = FALSE}
se = MakeSE(
    fst_path = "./NxtIRF_Output", 
    colData = colData
)
```

# Loading the Differential Expression Analysis

We will load the RDS file of the differential expression saved in the last vignette:

## Graphics user interface:

Navigate to the Analysis -> Differential Expression Analysis tab. Then click the "Load DE"
button at the bottom row of buttons. Select the previously saved RDS file containing the
differential expression. When the saved RDS file is loaded, it should look like this:

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Loading the Differential Expression Analysis"}
knitr::include_graphics("img/Vis_2_DE.png")
```

## Command line interface:

We will load the data frame containing our differential expression analysis as saved from
the last vignette:

```{r, eval = FALSE}
res.limma = data.table::fread("res.limma.csv", data.table = FALSE)
```

# Volcano Plots:

We will now draw a volcano plot. Volcano plots illustrate each differentially-expressed 
alternate splicing event (ASE) with the x axis denoting the log-2 fold change between 
treatment and control, and the y axis denoting p-value significance under negative
log-10 transformation. Significant ASE will be plotted higher than less significant ASEs.
Downregulated events will appear on the top-left, whereas upregulated events will appear
on the top-right. Non-differentially expressed events will appear on the bottom-center region.

## Graphics user interface:

Click "Display" on the navbar, then click "Volcano". The graph will be automatically generated
for the top 10k events of all types of alternative splicing:

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Volcano Plot"}
knitr::include_graphics("img/Vis_3_Volc_1.png")
```

To filter for the top events only, slide the "Number of Top Events" slider to the left (shown below). To include all events (not useful here since there are only around 1000 events), slide it all the way to the right.

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Volcano Plot of Top 100 events"}
knitr::include_graphics("img/Vis_3_Volc_2.png")
```

The volcano plot can also be subsetted by the type of alternative splicing. First, reset the top slider
to display 10000 events. Then, click the "Facet by Type" switch to turn it on.

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Volcano Plot by Type of Alternative Splicing"}
knitr::include_graphics("img/Vis_3_Volc_3.png")
```

## Command line interface:

We use the ggplot2 package to generate the equivalent plots. Interactive plots are not supported
using the command line interface.

```{r, eval = FALSE}
library(ggplot2)

ggplot(res.limma, aes(x = logFC, y = -log10(adj.P.Val))) + geom_point()
```

To facet by the type of alternative splicing:

```{r, eval = FALSE}
ggplot(res.limma, aes(x = logFC, y = -log10(adj.P.Val))) + geom_point() +
    facet_wrap(vars(EventType))
```

# Scatter Plots

Scatter (or diagonal) plots illustrate each ASE by comparing the fraction spliced in
or IR-ratios of alternative splicing and IR events, respectively. They contrast two
conditions with the x and y axes representing the average values of the two conditions.

## Graphics user interface:

Navigate to the "Display" -> "Diagonal". The canvas is blank for now as we have not set
which conditions to contrast.

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Scatter Plot settings"}
knitr::include_graphics("img/Vis_4_Diag_1.png")
```

To set our two conditions, click the drop-down box "Variable" and select "Treatment".
Next, select "D2" for "X-axis condition" and "UT" for "Y-axis condition". The
graph should now look like this:

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Scatter Plot"}
knitr::include_graphics("img/Vis_4_Diag_2.png")
```

## Command line interface:

The ggplot version of the scatter / diagonal plot:

```{r, eval = FALSE}
ggplot(res.limma, aes(x = AvgPSI_D2, y = AvgPSI_UT)) + geom_point()
```

# Heatmaps

Before we navigate to heatmaps, we should demonstrate how to select the top events using
interactive plots. First, navigate back to "Display" -> "Volcano". Next, select
"Facet by Type". Now, we can select the top significant events by type by using the mouse
to drawing a lasso or loop around the points we are interested in 
(just like gating populations on flow cytometry software). Below, we select the top 3 
alternative 3'-splice site "A3SS" events by drawing a lasso:

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Selecting Events"}
knitr::include_graphics("img/Vis_5_Gate_1.png")
```

Hold the mouse button to draw the lasso, and let go of the button to complete the selection.
The selected events will be coloured red. Draw a few lassos around the top 2 IR events, as well as
the top two skipped exons "SE" events. When we are done, it should look like this:

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Selected Top Events"}
knitr::include_graphics("img/Vis_5_Gate_2.png")
```

We can see these highlighted events on our diagonal plot, by going back to "Display" -> "Diagonal"

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Scatter Plot highlighting Selected Events"}
knitr::include_graphics("img/Vis_5_Gate_3.png")
```

Now, we can view these on our heatmap. Go to "Display" -> "Heatmap". The default figure should display
the selected events, coloured by their fraction spliced in (named here as PSI to denote the better-known
"Percent Spliced In"). However we keep the value as a fraction instead of a percentage.

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Heatmap of Selected Events"}
knitr::include_graphics("img/Vis_6_Heat_1.png")
```

Instead of our manually selected events, we can also graph the top N events in the entire list
(click on "Top N All Results") or from a list filtered by custom filters on the "Analysis" -> 
"Differential Expression Analysis" page (click on "Top N Filtered Results").

To change the palette colors, choose a different palette from the Palette drop-down menu.

PSI (fracton spliced in) gives the fraction of the included isoform divided by total transcripts.
This is a linear value. Sometimes it is more useful to use log-transformed values. Because alternative
splicing events often vary by log-scale at their extremes and by linear scale towards PSI ~ 0.5, we find
it useful to perform a logit-transformation (Mode = "Logit"), Alternatively, we can use Z-score transformation
of logit values ("Z-score")

To reset the selected events, navigate to "Analysis" -> "Differential Expression Analysis". You will notice that
our previously selected events are highlighted in deep blue:

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Unselecting Selected Events"}
knitr::include_graphics("img/Vis_7_Reset_1.png")
```

Click on the "Clear Selected Events" to clear all selected events.

## Command line interface:

NxtIRF uses the make_matrix() function to generate the requisite values behind the heatmap data.

To generate the PSI values of the top 20 ASE events:

```{r, eval = FALSE}
mat = make_matrix(
    se = se.filtered, 
    event_list = res.limma$EventName[1:20],
    method = "PSI"      # use "logit" for logit-transformed values, or "Z-score" for Z-score transformed values
)
```

Then, we use the pheatmap package to generate the heatmap. Feel free to use other heatmap packages.

```{r, eval = FALSE}
library("genefilter")
library("pheatmap")
library("RColorBrewer")

color = colorRampPalette(rev(brewer.pal(n = 5, name ="RdYlBu")))(100)

pheatmap(mat, color = color, breaks = seq(0, 1, length.out = 101),
    main = "Top 20 ASE Events in sample files")
```

# Coverage Plots

## Graphics User Interface:

Finally, navigate to the "Display" -> "Coverage". The Genes and Events drop-down boxes
take some time to load, but once they do they will be filled with Genes and ASE events
(the latter in order of significance from our differential analysis)

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Coverage Plot Tab - Selecting Events"}
knitr::include_graphics("img/Vis_8_Cov_1.png")
```

Just like heatmaps, you can select to display all top events (the number is controlled 
by a slider on the right hand side), or by filtered or user-highlighted events.

Now, select the top event which is "ANAPC2 Intron 4" using the "Events" drop-down box.

Then, on the left hand panel (just to the right of the navigation bar), notice that "View" is set to
"Individual". This means individual samples will be displayed, each in its own track, just like in
the Integrated Genomics Viewer (IGV). Select "UT_1" for Track 1, "D2_1" for Track 2, then click "Refresh Plot"
which is the button just above the column of drop-down boxes (highlighted by the red arrow below):

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Setting Tracks and Generating Plots"}
knitr::include_graphics("img/Vis_8_Cov_2.png")
```

Once you click "Refresh Plot", the plot will be generated below:

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Coverage Plot of Individual Tracks"}
knitr::include_graphics("img/Vis_8_Cov_3.png")
```

## Generating Coverage Plots by Condition

Now, we are only comparing one sample to another. NxtIRF provides a tool to combine
all the coverage plots per condition, to allow us to compare the coverage between two
conditions. This is possible because NxtIRF first normalises the coverages based on their
transcript abundance using the "SpliceOver" metric.

To plot the ANAPC2 intron 4 IR event by condition, change the "View" to "Condition". A new
drop-down box called "Condition" will appear. Set this to "Treatment", then set "UT" 
for Track 1 and "D2" for Track 2. Then, press "Refresh Plot":

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Coverage Plot by Condition"}
knitr::include_graphics("img/Vis_8_Cov_4.png")
```

You will notice that there are fuzzy regions around the coverage line. This is the 95% confidence
interval calculated from all the samples belonging to the conditions.

To compare the two conditions in a single plot, click on the "Stack Traces" switches
which is one of the switches below the drop-down boxes for the 4 tracks. Then click "Refresh Plot"

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Stacked Coverage Plot"}
knitr::include_graphics("img/Vis_8_Cov_5.png")
```

To plot the pair-wise T test comparing the coverages of the two conditions, per-nucleotide,
click the "Pairwise t-test" switch to turn it on. Then, click "Refresh Plot":

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Stacked Coverage Plot with T-test"}
knitr::include_graphics("img/Vis_8_Cov_6.png")
```

To view this intron in the context of the whole gene, go to the "Genes" drop-down box at the top right,
type in "ANAPC2", then select the gene. Then, press "Refresh Plot"

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Plot by Gene"}
knitr::include_graphics("img/Vis_8_Cov_7.png")
```

Note that there are multiple introns retained in this gene. However it is difficult to tell,
because the normalisation at intron 4 does not lead to proper normalisation of the downstream gene
(on the left, as ANAPC2 is reversely stranded). To normalise to a different intron, go to the
drop-down box "Normalise Event" (the 2nd from "Refresh Plot"), and select "ANAPC2 Intron 12".
Note you may have to resize the app's window to see the full text. When you have done so, click
"Refresh Plot".

```{r, echo=FALSE, out.width='640pt', fig.align = 'center', fig.cap="Normalise by 3-prime most intron"}
knitr::include_graphics("img/Vis_8_Cov_8.png")
```

Now, the downstream gene (left) is well-normalised, but the upstream gene is not well-normalised.
It is important to normalise to the IR or splicing event you are evaluating. This is because
differences in 3-prime bias may affect proper normalisation of ASE events. Additionally,
retained introns lead to longer transcripts which compound 3-prime bias.

Feel free to play with the other controls. Please note that this is a beta version of the software
so there may be potential bugs.

## Command line interface:

First, we need to load the transcript, exon and intron data from the NxtIRF reference. To load data
for use in Coverage plots, use the following:

```{r, eval = FALSE}
cov_data = prepare_covplot_data("./Reference")
```

To plot ANAPC2-Intron 4 (the top event), using the individual samples "UT_1" and "D2_1"

```{r, eval = FALSE}
i = 1
pl = Plot_Coverage(
    se = se, 
    Event = res.limma$EventName[i], 
    cov_data = cov_data, 
    tracks = c("UT_1", "D2_1"), 
    )
```

The Plot_Coverage() function returns a list of two objects. These are the ggplot and plotly
objects named as "ggplot" and "final_plot", respectively. To view the plotly version:

```{r, eval = FALSE}
pl$final_plot
```

To view the ggplot version:

```{r, eval = FALSE}
egg::ggarrange(pl$ggplot[[1]], pl$ggplot[[2]], pl$ggplot[[6]], ncol = 1)
```

Note that ggplot[[1]] and ggplot[[2]] correspond to tracks 1 and 2, respectively. NxtIRF supports up to
4 tracks. Track 5 is reserved for Pairwise T test plot (not used here), and Track 6 is the gene annotation.

To plot by conditions "UT" and "D2", with all the coverage traces on one track, and the T-test on another:

```{r, eval = FALSE}
i = 1
pl = Plot_Coverage(
    se = se, 
    Event = res.limma$EventName[i], 
    cov_data = cov_data, 
    tracks = c("UT", "D2"),
    condition = "Treatment",
    stack_tracks = TRUE,
    t_test = TRUE
    )
```

Type in `pl$final_plot` to view the plotly version. For ggplot:

```{r, eval = FALSE}
egg::ggarrange(pl$ggplot[[1]], pl$ggplot[[5]], pl$ggplot[[6]], ncol = 1)
```

All the coverage traces are now in Track 1. The t-test track is on Track 5.

To view the intron by the whole ANAPC2 gene, only displaying reads belonging to
the negative strand:

```{r, eval = FALSE}
i = 1
pl = Plot_Coverage(
    se = se,
    Event = res.limma$EventName[i], 
    cov_data = cov_data, 
    tracks = c("D2", "UT"), 
    condition = "Treatment", 
    stack_tracks = TRUE, t_test = TRUE,
    Gene = "ANAPC2", 
    zoom_factor = 0, 
    strand = "-")
pl$final_plot
```

Here, `zoom_factor` is used in NxtIRF to zoom out from the boundaries as defined by the parameters.
Here, these are the boundaries of ANAPC2 (as `Gene` is defined). Each level of zoom
will zoom out by a factor of 3. So, to zoom out, try setting zoom_factor = 1. NB negative zoom
values are not allowed.

Also, notice that the normalisation is set to normalise at Intron 4. This is because this is given
by setting Event as the top event (res.limma$EventName[i], where i = 1).

To normalise for intron 12, simply replace Event with the name for Intron 12. To search for this,
simply do this:

```{r, eval = FALSE}
res.limma$EventName[which(grepl("ANAPC2", res.limma$EventName))]
```

After displaying all the names of the events associated with ANAPC2, do the following:

```{r, eval = FALSE}
pl = Plot_Coverage(
    se = se,
    Event = "ANAPC2/ENST00000323927_Intron12/clean", 
    cov_data = cov_data, 
    tracks = c("D2", "UT"), 
    condition = "Treatment", 
    stack_tracks = TRUE, t_test = TRUE,
    Gene = "ANAPC2", 
    zoom_factor = 0, 
    strand = "-")
pl$final_plot
```
